Upgrade SDK v2

amagana3 opened the pull request:
### Overview
In preparation for the new and improved [Terraform 1.0.0 release](https://www.hashicorp.com/blog/announcing-hashicorp-terraform-1-0-general-availability), we've gone ahead and shaped up the ALKS Terraform Provider to use the new SDK v2. With the help of this amazing guide: https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html, all functionality remains practically the same.

### Automatic Changes
Most of the changes are actually just changes to the packages used. Previously, we were actually _not_ using the Terraform SDK, but the Terraform platform itself to make calls. While this wasn't necessarily _wrong_ or _incorrect_, it wasn't the proper way to configure our custom provider. Instead, we now use the following standalone SDK: https://github.com/hashicorp/terraform-plugin-sdk, so this should be more lightweight. 

With the help of the handy [tf-sdk-migrator](https://github.com/hashicorp/tf-sdk-migrator), I was able to use the following commands to make this upgrade more seamless:
- `tf-sdk-migrator migrate`
   - This automatically migrates the Terraform provider to the new extracted SDK (github.com/hashicorp/terraform-plugin-sdk), replacing references to the old SDK (github.com/hashicorp/terraform).
- `tf-sdk-migrator v2upgrade`
   - This automatically migrates a Terraform provider using version 1.x of the standalone SDK to version 2.x of the standalone SDK, updating package import paths.

### Manual Changes
Most of the manual changes were done by referencing the following document: https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html. Mostly deprecated methods, switching up our tests, but most important re-configuring our `provider.go` file to use the new and improved definitions. Here's a snippet of one of the many changes made in this refactor:
> The terraform.ResourceProvider interface has been removed. The concrete *helper/schema.Provider type should be used in its place. For most providers, this will involve changing the returned type of the Provider() function, and removing some type assertions; all necessary changes should be raised at compile time.

### Testing
All functionality was tested which include: 
- Creating an ALKS IAM Role
- Creating an ALKS IAM Trust Role
- Creating an LTK
- Deleting all of the above.
- Creating all with multi-provider configurations in different accounts.

Pretty much all functionality remained the same, and all the unit tests also passed.

## Disclaimer
This is going to be a **major** version bump as recommended by Terraform. See the following quote:
> It is recommended, but not required, that providers that have not already dropped support for 0.11 issue a major version bump when upgrading to version 2 of the SDK, to indicate this dropped support.

So yes, 0.11 support will be **dropped** after this release. But users are still able to use any previous version of our custom provider to use in their 0.11 projects.

Good luck reviewer! 👍🏼

rballard-cai commented:
Ran into an error when creating a cross account role. Strangely it actually does create the role.

```
terraform {
  required_providers {
    alks = {
      source  = "Cox-Automotive/engineering-enablement/alks"
      version = "2.0.0"
    }
    aws = {
      source = "hashicorp/aws"
    }
  }
  required_version = ">= 1.0.1"
}

provider "alks" {
  url     = "https://alks.coxautoinc.com/rest"
  profile = "awsmarvennp"
}

resource "alks_iamrole" "test_role" {
  name                     = "ryan-test-role"
  type                     = "Amazon EC2"
  include_default_policies = false
  enable_alks_access       = true
}

resource "alks_iamtrustrole" "test_trust_role" {
  name               = "ryan-test-cross-role"
  type               = "Cross Account"
  trust_arn          = "arn:aws:iam::464247036087:role/acct-managed/ryan-test-cross-account"
  enable_alks_access = false
}

resource "alks_ltk" "test_ltk_user" {
  iam_username = "ryan-test-ltk-user"
}
```

```
Terraform will perform the following actions:

  # alks_iamrole.test_role will be created
  + resource "alks_iamrole" "test_role" {
      + arn                      = (known after apply)
      + enable_alks_access       = true
      + id                       = (known after apply)
      + include_default_policies = false
      + ip_arn                   = (known after apply)
      + name                     = "ryan-test-role"
      + role_added_to_ip         = (known after apply)
      + type                     = "Amazon EC2"
    }

  # alks_iamtrustrole.test_trust_role will be created
  + resource "alks_iamtrustrole" "test_trust_role" {
      + arn                = (known after apply)
      + enable_alks_access = false
      + id                 = (known after apply)
      + ip_arn             = (known after apply)
      + name               = "ryan-test-cross-role"
      + role_added_to_ip   = (known after apply)
      + trust_arn          = "arn:aws:iam::464247036087:role/acct-managed/ryan-test-cross-account"
      + type               = "Cross Account"
    }

  # alks_ltk.test_ltk_user will be created
  + resource "alks_ltk" "test_ltk_user" {
      + access_key   = (sensitive value)
      + iam_user_arn = (known after apply)
      + iam_username = "ryan-test-ltk-user"
      + id           = (known after apply)
      + secret_key   = (sensitive value)
    }

Plan: 3 to add, 0 to change, 0 to destroy.
```

```
alks_iamtrustrole.test_trust_role: Creating...
╷
│ Error: empty non-retryable error received. This is a bug with the Terraform provider and should be reported as a GitHub issue in the provider repository.
│
│   with alks_iamtrustrole.test_trust_role,
│   on main.tf line 13, in resource "alks_iamtrustrole" "test_trust_role":
│   13: resource "alks_iamtrustrole" "test_trust_role" {
```

rballard-cai commented:
Ran into an error when creating a cross account role. Strangely it actually does create the role.

```
terraform {
  required_providers {
    alks = {
      source  = "Cox-Automotive/engineering-enablement/alks"
      version = "2.0.0"
    }
    aws = {
      source = "hashicorp/aws"
    }
  }
  required_version = ">= 1.0.1"
}

provider "alks" {
  url     = "https://<url>"
  profile = "<account>"
}

resource "alks_iamrole" "test_role" {
  name                     = "ryan-test-role"
  type                     = "Amazon EC2"
  include_default_policies = false
  enable_alks_access       = true
}

resource "alks_iamtrustrole" "test_trust_role" {
  name               = "ryan-test-cross-role"
  type               = "Cross Account"
  trust_arn          = "arn:aws:iam::12345780:role/acct-managed/ryan-test-cross-account"
  enable_alks_access = false
}

resource "alks_ltk" "test_ltk_user" {
  iam_username = "ryan-test-ltk-user"
}
```

```
Terraform will perform the following actions:

  # alks_iamrole.test_role will be created
  + resource "alks_iamrole" "test_role" {
      + arn                      = (known after apply)
      + enable_alks_access       = true
      + id                       = (known after apply)
      + include_default_policies = false
      + ip_arn                   = (known after apply)
      + name                     = "ryan-test-role"
      + role_added_to_ip         = (known after apply)
      + type                     = "Amazon EC2"
    }

  # alks_iamtrustrole.test_trust_role will be created
  + resource "alks_iamtrustrole" "test_trust_role" {
      + arn                = (known after apply)
      + enable_alks_access = false
      + id                 = (known after apply)
      + ip_arn             = (known after apply)
      + name               = "ryan-test-cross-role"
      + role_added_to_ip   = (known after apply)
      + trust_arn          = "arn:aws:iam::123457890:role/acct-managed/ryan-test-cross-account"
      + type               = "Cross Account"
    }

  # alks_ltk.test_ltk_user will be created
  + resource "alks_ltk" "test_ltk_user" {
      + access_key   = (sensitive value)
      + iam_user_arn = (known after apply)
      + iam_username = "ryan-test-ltk-user"
      + id           = (known after apply)
      + secret_key   = (sensitive value)
    }

Plan: 3 to add, 0 to change, 0 to destroy.
```

```
alks_iamtrustrole.test_trust_role: Creating...
╷
│ Error: empty non-retryable error received. This is a bug with the Terraform provider and should be reported as a GitHub issue in the provider repository.
│
│   with alks_iamtrustrole.test_trust_role,
│   on main.tf line 13, in resource "alks_iamtrustrole" "test_trust_role":
│   13: resource "alks_iamtrustrole" "test_trust_role" {
```

amagana3 commented:
Great find @rballard-cai 

You can see the Retry was modified and changed in the new SDK here: https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#clearer-handling-of-nil-for-helper-resource-nonretryableerror-and-helper-resource-retryableerror

So I've gone ahead and modified the return to match what the behavior was before. The next PR when refactoring all deprecated methods I will modify this retry logic to the new and improved version as seen here: https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#clearer-handling-of-nil-for-helper-resource-nonretryableerror-and-helper-resource-retryableerror

rballard-cai commented on resource_alks_iamrole.go:
```diff
@@ -179,9 +178,10 @@ func resourceAlksIamTrustRoleCreate(d *schema.ResourceData, meta interface{}) er
+		return nil
 	})
 
+	//resp, err := client.CreateIamTrustRole(roleName, roleType, trustArn, enableAlksAccess)
```
forgot to delete this?

amagana3 replied:
Good catch - deleted!

rballard-cai commented:
Tested successfully using the following

```
provider "alks" {
  url     = "..."
  profile = "..."
}

resource "alks_iamrole" "test_role" {
  name                     = "ryan-test-role"
  type                     = "Amazon EC2"
  include_default_policies = false
  enable_alks_access       = true
}

resource "alks_iamtrustrole" "test_trust_role" {
  name               = "ryan-test-cross-role"
  type               = "Cross Account"
  trust_arn          = "arn:aws:iam::...:role/acct-managed/ryan-test-cross-account"
  enable_alks_access = false
}

resource "alks_iamtrustrole" "test_trust_inner_role" {
  name               = "ryan-test-inner-role"
  type               = "Inner Account"
  trust_arn          = alks_iamrole.test_role.arn
  enable_alks_access = false
}

resource "alks_ltk" "test_ltk_user" {
  iam_username = "ryan-test-ltk-user"
}
```

rballard-cai commented on .gitignore:
```diff
@@ -14,3 +14,6 @@ glide.lock
 gpg/
+
+# ignore. For testing we don't want to store a specific provider.
+.terraform.lock.hcl
```
Should this still be ignored?

amagana3 replied:
Yes, if we keep this then whenever the provider is changed (since locally one can download the binary) then the lock file will keep an old version cached and throw errors. To prevent this, we should ignore the lock file

rballard-cai commented on docs/guides/local_installation.md:
```diff
@@ -93,6 +93,6 @@ mkdir -p ~/.terraform.d/plugins/Cox-Automotive/engineering-enablement/alks/1.5.1
-| Terraform 0.10.x       | Terraform 0.12.x  | Terraform 0.13.x | Terraform 0.14.x |
-| ----------------       | ----------------  | ---------------- | ---------------- |
-| ALKS TFP 0.9.0 < 1.3.0 | ALKS TFP 1.3.0+   | ALKS TFP 1.3.0+  | ALKS TFP 1.3.0+  |
+| Terraform 0.10.x       | Terraform 0.12.x  | Terraform 0.13.x | Terraform 0.14.x | Terraform 0.11.x |
```
Could you move 0.11.x to inbetween 0.10 and 0.12?

amagana3 replied:
Ah yeah committed this incorrectly, good catch
